README.md

[![Bonsai Asset Badge](https://img.shields.io/badge/Sensu%20AWS-Download%20Me-brightgreen.svg?colorB=89C967&logo=sensu)](https://bonsai.sensu.io/assets/sensu/sensu-aws)

# Sensu Go AWS Plugin Collection

- [Overview](#overview)
- [Usage examples](#usage-examples)
  - [Golang program locations](#golang-program-locations)
- [Configuration](#configuration)
  - [Asset registration](#asset-registration)
  - [Check definition](#check-manifest)
  - [On-disk configuration](#on-disk-configuration)
- [Installation from source](#installation-from-source)
- [Contributing](#contributing)

## Overview

This Sensu Go plugin collection provides metric and status checks for monitoring AWS services with Sensu. 

## Usage examples

**check-alb-target-group-health**

```
  ./check-alb-target-group-health --aws_region=us-east-1

  ./check-alb-target-group-health --aws_region=us-east-1 --target_groups=target-group-1
  
  ./check-alb-target-group-health --aws_region=us-east-1 --target_groups=target-group-a,target-group-b
  
```
**check-cloudwatch-alarm**

```
  ./check-cloudwatch-alarm --aws_region=eu-west-1
   
  ./check-cloudwatch-alarm --state=ALEARM

```
**check-cloudwatch-alarms**

```
  ./check-cloudwatch-alarms --exclude_alarms=CPUAlarmLow
  
  ./check-cloudwatch-alarms --aws_region=eu-west-1 --exclude_alarms=CPUAlarmLow
  
  ./check-cloudwatch-alarms --state=ALEARM
      
```
**check-cloudwatch-composite-metric**

```
  ./check-cloudwatch-composite-metric --namespace AWS/ELB --dimensions="LoadBalancerName=test-elb" 
                                      --period=60 --statistics=Maximum --operator=equal --critical=0 
```

**check-ebs-burst-limit**

```
  ./check-ebs-burst-limit --aws_region=eu-west-1
  
```

**check-ebs-snapshots**

```
  ./check-ebs-snapshots --check_ignored=false
  
```

**check-ec2-cpu_balance**

```
  ./check-ec2-cpu_balance --critical=3
  
  ./check-ec2-cpu_balance --critical=1 --warning=5
  
  ./check-ec2-cpu_balance --critical=1 --warning=5 --tag=TESTING
  
```

**check-ec2-filter**

```
  ./check-ec2-filter --filters="{\"filters\" : [{\"name\" : \"instance-state-name\", \"values\": [\"running\"]}]}"
  
  ./check-ec2-filter --exclude_tags="{\"TAG_NAME\" : \"TAG_VALUE\"}" --compare=not
  
```

**check-ec2-network**

```
  ./check-ec2-network --instance_id=i-0f1626fsbfvbafa2 --direction=NetworkOut
  
```

**metrics-ec2-count**

```
  ./metrics-ec2-count --metric_type=status
  
  ./metrics-ec2-count --metric_type=instance
  
```

**metrics-ec2-filter**

```
  ./metric-ec2-filter --filters="{\"filters\" : [{\"name\" : \"instance-state-name\", \"values\": [\"running\"]}]}"
  
```

**check-elb-certs**

```
  ./check-elb-certs --aws_region=${your_region} --warning=${days_to_warn} -critical=${days_to_critical}
  
```

**check-elb-health-fog**

```
  ./check-elb-health-fog --aws_region=${you_region} --instances=${your_instance_ids} 
                         --elb_name=${your_elb_name} --verbose=true  
```

**check-elb-health-sdk**

```
  ./check-elb-health-sdk --aws_region=region
  
  ./check-elb-health-sdk --aws_region=region --elb_name=my-elb
  
  ./check-elb-health-sdk --aws_region=region --elb_name=my-elb --instances=instance1,instance2
  
```

**check-elb-instance-inservice**

```
  ./check-elb-instance-inservice --aws_region=${your_region}
  
  ./check-elb-instance-inservice --aws_region=${your_region} --elb_name=${LoadBalancerName}
  
```

**check-elb-latency**

```
  ./check-elb-latency --warning_over=1 --critical_over=3
  
  ./check-elb-latency --elb_names=app --critical_over=5 --statistics=maximum --period=3600
  
```

**check-elb-nodes**

```
  ./check-elb-nodes --warning=3 --critical=2 --load_balancer=#{your-load-balancer}

  ./check-elb-nodes --warning_percentage=50 --critical_percentage=25 --load_balancer=#{your-load-balancer}
  
```

**elb-metrics**

```
  ./elb-metrics --aws_region=${your_region}
  
```

**check-rds**

```
  ./check-rds --db_instance_id=sensu-admin-db --available_zone_severity=critical --available_zone=ap-northeast-1a
  
  ./check-rds --db_instance_id=sensu-admin-db --cpu_warning_over=80 --cpu_critical_over=90
  
  ./check-rds --db_instance_id=sensu-admin-db --cpu_critical_over=90 --statistics=maximum --period=3600
  
  ./check-rds --db_instance_id=sensu-admin-db --connections_critical_over=120 --connections_warning_over=100 
              --statistics=maximum --period=3600
              
  ./check-rds --db_instance_id=sensu-admin-db --iops_critical_over=200 --iops_warning_over=100 --period=300
  
  ./check-rds --db_instance_id=sensu-admin-db --memory_warning_over=80 --statistics=minimum --period=7200
  
  ./check-rds --db_instance_id=sensu-admin-db --disk_warning_over=80 --period=7200
  
  ./check-rds --db_instance_id=sensu-admin-db --cpu_warning_over=80 --cpu_critical_over=90 
              --memory_warning_over=60 --memory_critical_over=80
              
 ```
 
 **check-rds-events**
 
 ```
  ./check-rds-events --aws_region=${your_region}  --db_instance_id=${your_rds_instance_id_name}
  
  ./check-rds-events.rb --aws_region=${your_region}
  
  ```
  
 **check-rds-pending**
  
 ```
   ./check-rds-pending --aws_region=${you_region}

 ```

 **rds-metrics**
 
 ```
  ./rds-metrics --aws_region=eu-west-1
  
  ./rds-metrics --aws_region=eu-west-1 --db_instance_id=sr2x8pbti0eon1
  
 ```
 
 **check-s3-bucket**
 
 ```
  ./check-s3-bucket --bucket_name=mybucket
  
 ```
 
 **check-s3-bucket-visibility**
 
 ```
  ./check-s3-bucket-visibility.go --exclude_buckets_regx=sensu --bucket_names=ssensu-ec2,sensu-ec3 
                                  --exclude_cuckets=sensu-ec3
 ```
 
 **check-s3-object**
 
 ```
  ./check-s3-object --bucket_name=aws-testing --key_prefix=s3
  
 ```
  
 **check-s3-tag**
 
 ```
  ./check-s3-tag --tag_keys=sensu
  
 ```
 
 **s3-metrics**
 
 ```
  ./metrics-s3
  
 ```
 
 ### Golang program locations
 
 You can find the individual Golang programs in https://github.com/sensu/sensu-aws/tree/master/plugins. In that directory you'll find a listing of AWS services, with each service directory housing its respective checks.  

## Configuration

### Asset registration

Assets are the best way to make use of this handler. If you're not using an asset, please consider doing so! If you're using Sensu 5.13 or later, you can use the following command to add the asset: 

`sensuctl asset add sensu/sensu-aws`

If you're using an earlier version of Sensu, you can download the asset definition from [this project's Bonsai Asset Index page](https://bonsai.sensu.io/assets/sensu/sensu-aws).

### Check definition

```yaml
---
type: CheckConfig
api_version: core/v2
metadata:
  name: check-ec2-network
  namespace: default
spec:
  check_hooks: null
  command: check-ec2-network -instance_id i-1234567890
  env_vars: null
  handlers: []
  high_flap_threshold: 0
  interval: 10
  low_flap_threshold: 0
  output_metric_format: ""
  output_metric_handlers: null
  proxy_entity_name: ""
  publish: true
  round_robin: false
  runtime_assets:
  - sensu/sensu-aws
  stdin: false
  subdue: null
  subscriptions:
  - aws
  timeout: 0
  ttl: 0
```

### On-disk configuration

This plugin requires on-disk configuration for any host that will run checks from this collection. See the Sample Credential Configuration below to create the file you need to operate this plugin.

```
Sample Credential Configuration:
[default]
aws_access_key_id=AGFGFHGFHHJGHJG
aws_secret_access_key=cdfedbfdjdjsbsjdgbjdsgbjdskg

Sample Config:
[default]
region=us-east-2
output=json

$ mkdir ~/.aws
$ cd ~/.aws
$ vi credentials - copy and paste the above sample credential and change the aws_access_key_id and aws_secret_access_key to some valid values. Save the file.
$ vi config - copy and paste the above sample config and change the region to some valid value. Save the file.
```

## Installation from source

The preferred way to install and deploy this plugin is to use it as an [asset][2]. To compile and install the plugin from source or contribute to the plugin, download the latest version of the sensu-aws from [releases][1] or create an executable script from this source.

From the local path of the sensu-aws repository:

```
go build -o /usr/local/bin/sensu-aws main.go
```
## Contributing
For more information about contributing to this plugin, see https://github.com/sensu/sensu-go/blob/master/CONTRIBUTING.md

[1]: https://github.com/sensu/sensu-aws/releases
[2]: #asset-registration


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic
Versioning](http://semver.org/spec/v2.0.0.html).

## Unreleased

## [0.0.0] - 2020-09-08

### Changed
- Sync up gorelaser and bonsai releases

## [0.0.8] - 2020-09-08

### Changed
- Replaced Travis with GitHub Actions

## [0.0.7] - 2020-09-08

### Changed
- Fixed goreleaser for archives instead of archivev

## [0.0.6] - 2020-09-08

### Changed 
- Updated goreleaser targets
- Fixed rds start time using fetch age for calculating duration
- Clean up check-alb-target-group-health

## [0.0.3] - 2019-12-05

### Changed 

- Address goreleaser issues re: id's
- Update Travis config & move to Go modules
- Updated README.md 

## [0.0.2] - 2019-01-18

### Added
- Added Bonsai YAML config file

## [0.0.1] - 2019-01-18

### Added
- Initial release


plugins/rds/check-rds/main.go
package main

/*
#
# check-rds
#
# DESCRIPTION:
#   Check RDS instance statuses by RDS and CloudWatch API.
#
# OUTPUT:
#   plain-text
#
# PLATFORMS:
#   MAC OS
#
#
# USAGE:
#   Critical if DB instance "sensu-admin-db" is not on ap-northeast-1a
#   ./check-rds --db_instance_id=sensu-admin-db --available_zone_severity=critical --available_zone=ap-northeast-1a
#
#   Warning if CPUUtilization is over 80%, critical if over 90%
#   ./check-rds --db_instance_id=sensu-admin-db --cpu_warning_over=80 --cpu_critical_over=90
#
#   Critical if CPUUtilization is over 90%, maximum of last one hour
#   ./check-rds --db_instance_id=sensu-admin-db --cpu_critical_over=90 --statistics=maximum --period=3600
#
#   Warning if DatabaseConnections are over 100, critical over 120
#   ./check-rds --db_instance_id=sensu-admin-db --connections_critical_over=120 --connections_warning_over=100 --statistics=maximum --period=3600
#
#   Warning if IOPS are over 100, critical over 200
#   ./check-rds --db_instance_id=sensu-admin-db --iops_critical_over=200 --iops_warning_over=100 --period=300
#
#   Warning if memory usage is over 80%, maximum of last 2 hour
#   specifying "minimum" is intended actually since memory usage is calculated from CloudWatch "FreeableMemory" metric.
#   ./check-rds --db_instance_id=sensu-admin-db --memory_warning_over=80 --statistics=minimum --period=7200
#
#   Disk usage, same as memory
#   ./check-rds --db_instance_id=sensu-admin-db --disk_warning_over=80 --period=7200
#
#   You can check multiple metrics simultaneously. Highest severity will be reported
#   ./check-rds --db_instance_id=sensu-admin-db --cpu_warning_over=80 --cpu_critical_over=90 --memory_warning_over=60 --memory_critical_over=80
#
#   You can ignore accept nil values returned for a time periods from Cloudwatch as being an OK.  Amazon falls behind in their
#   metrics from time to time and this prevents false positives
#   ./check-rds --db_instance_id=sensu-admin-db --cpu_critical_over=90 -accept_nil=true
#
# NOTES:
#
# LICENSE:
#   TODO
#
*/

import (
	"fmt"
	"math"
	"os"
	"strings"
	"time"

	"github.com/sensu/sensu-aws/awsclient"
	"github.com/spf13/cobra"

	"github.com/aws/aws-sdk-go/aws"

	"github.com/aws/aws-sdk-go/service/cloudwatch"
	"github.com/aws/aws-sdk-go/service/rds"
	"github.com/sensu/sensu-aws/aws_session"
)

var (
	awsRegion                string
	rdsClient                *rds.RDS
	scheme                   string
	dbInstanceId             string
	fetchAge                 int
	period                   int64
	statistic                string
	cloudWatchClient         *cloudwatch.CloudWatch
	roleArn                  string
	dbClusterId              string
	accpetNil                bool
	availabilityZoneSeverity string
	cpuCritical              float64
	cpuWarning               float64
	memoryCritical           float64
	memoryWarning            float64
	diskCritical             float64
	diskWarning              float64
	conectionCritical        float64
	connectionWarning        float64
	iopsCritical             float64
	iopsWarning              float64
	metricSeverities         map[string]map[string]float64
	availabilityZone         string
	instanceClassMapping     map[string]string
	allocatedStorageMapping  map[string]int64
	dbInstanceZoneMapping    map[string]string
)

func checkRds() {
	var success bool
	metrics := getMetrics()
	metricSeverities = getMetricSeverities()
	values := make(map[string]*float64)
	severities := make(map[string]bool)
	severities["critical"] = false
	severities["warning"] = false
	instanceClassMapping = make(map[string]string)
	allocatedStorageMapping = make(map[string]int64)
	dbInstanceZoneMapping = make(map[string]string)
	if len(dbClusterId) <= 0 && len(dbInstanceId) <= 0 {
		fmt.Println("Please provide db_cluster_id or db_instance_id")
		return
	}
	awsSession := aws_session.CreateAwsSessionWithRegion(awsRegion)
	if len(roleArn) <= 0 {
		success, rdsClient = awsclient.GetRDSClient(awsSession)
	} else {
		success, rdsClient = awsclient.GetRDSClientWithRoleArn(awsSession, roleArn)
	}
	if !success {
		return
	}
	if len(roleArn) <= 0 {
		success, cloudWatchClient = awsclient.GetCloudWatchClient(awsSession)
	} else {
		success, cloudWatchClient = awsclient.GetCloudWatchClientWithRoleArn(awsSession, roleArn)
	}
	if !success {
		return
	}
	err := getClusterDetails()
	if err != nil {
		return
	}
	err = getDbInstanceDetails()
	if err != nil {
		return
	}
	for instance, zone := range dbInstanceZoneMapping {
		//if zone != availabilityZone {
		// if availabilityZoneSeverity == "citical" {
		// 	fmt.Print("CRITICAL :")
		// } else {
		// 	fmt.Print("WARNING :")
		// }
		fmt.Println("Availabilty Zone for DB Instance", instance, "is", zone)
		//}
		for metric, unit := range metrics {
			value, err := getCloudWatchMetrics(metric, instance, unit)
			if err != nil {
				fmt.Println("Error :", err)
				return
			}
			values[metric] = value
		}

		checkCPU(values["CPUUtilization"], instance)
		checkMemory(values["FreeableMemory"], instance, instanceClassMapping[instance])
		checkDiskSpace(values["FreeStorageSpace"], instance, allocatedStorageMapping[instance])
		checkConnections(values["DatabaseConnections"], instance)
		checkIops(values["ReadIOPS"], values["WriteIOPS"], instance)
	}
}

func getCloudWatchMetrics(metricname string, rdsName string, unit string) (*float64, error) {
	var input cloudwatch.GetMetricStatisticsInput
	input.Namespace = aws.String("AWS/RDS")
	input.MetricName = aws.String(metricname)
	var dimensionFilter cloudwatch.Dimension
	dimensionFilter.Name = aws.String("DBInstanceIdentifier")
	dimensionFilter.Value = aws.String(rdsName)
	input.Dimensions = []*cloudwatch.Dimension{&dimensionFilter}
	input.EndTime = aws.Time(time.Now())
	input.StartTime = aws.Time((input.EndTime).Add(time.Duration(-fetchAge/60) * time.Minute))
	input.Period = aws.Int64(period)
	input.Statistics = []*string{aws.String(strings.Title(statistic))}
	input.Unit = aws.String(unit)
	metrics, err := cloudWatchClient.GetMetricStatistics(&input)
	if err != nil {
		return nil, err
	}
	if metrics != nil && metrics.Datapoints != nil && len(metrics.Datapoints) > 1 {
		var minimumTimeDifference float64
		var timeDifference float64
		var value *float64
		minimumTimeDifference = -1
		for _, datapoint := range metrics.Datapoints {
			timeDifference = time.Since(*datapoint.Timestamp).Seconds()
			if minimumTimeDifference == -1 || timeDifference < minimumTimeDifference {
				minimumTimeDifference = timeDifference
				if strings.Title(statistic) == "Average" {
					value = datapoint.Average
				} else if strings.Title(statistic) == "Sum" {
					value = datapoint.Sum
				} else if strings.Title(statistic) == "Maximum" {
					value = datapoint.Maximum
				} else if strings.Title(statistic) == "Minimum" {
					value = datapoint.Minimum
				}
			}
		}
		return value, nil
	}
	return nil, nil
}

func getMetrics() map[string]string {
	metrics := make(map[string]string)
	metrics["CPUUtilization"] = "Percent"
	metrics["FreeableMemory"] = "Bytes"
	metrics["FreeStorageSpace"] = "Bytes"
	metrics["DatabaseConnections"] = "Count"
	metrics["ReadIOPS"] = "Count/Second"
	metrics["WriteIOPS"] = "Count/Second"
	return metrics
}

func getMetricSeverities() map[string]map[string]float64 {
	metricSeverities := make(map[string]map[string]float64)
	metricSeverities["CPUUtilization"] = make(map[string]float64)
	metricSeverities["CPUUtilization"]["critical"] = cpuCritical
	metricSeverities["CPUUtilization"]["warning"] = cpuWarning
	metricSeverities["FreeableMemory"] = make(map[string]float64)
	metricSeverities["FreeableMemory"]["critical"] = memoryCritical
	metricSeverities["FreeableMemory"]["warning"] = memoryWarning
	metricSeverities["FreeStorageSpace"] = make(map[string]float64)
	metricSeverities["FreeStorageSpace"]["critical"] = diskCritical
	metricSeverities["FreeStorageSpace"]["warning"] = diskWarning
	metricSeverities["DatabaseConnections"] = make(map[string]float64)
	metricSeverities["DatabaseConnections"]["critical"] = conectionCritical
	metricSeverities["DatabaseConnections"]["warning"] = connectionWarning
	metricSeverities["ReadIOPS"] = make(map[string]float64)
	metricSeverities["ReadIOPS"]["critical"] = iopsCritical
	metricSeverities["ReadIOPS"]["warning"] = iopsWarning
	metricSeverities["WriteIOPS"] = make(map[string]float64)
	metricSeverities["WriteIOPS"]["critical"] = iopsCritical
	metricSeverities["WriteIOPS"]["warning"] = iopsWarning
	return metricSeverities
}

func configureRootCommand() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "check-rds",
		Short: "The Sensu Go Aws RDS handler for rds management",
		RunE:  run,
	}

	cmd.Flags().StringVar(&awsRegion, "aws_region", "us-west-1", "AWS Region (defaults to us-east-1).")
	cmd.Flags().StringVar(&scheme, "scheme", "", "Metric naming scheme, text to prepend to metric")
	cmd.Flags().StringVar(&dbInstanceId, "db_instance_id", "", "DB instance identifier")
	cmd.Flags().IntVar(&fetchAge, "fetch_age", 0, "How long ago to fetch metrics from in seconds")
	cmd.Flags().Int64Var(&period, "period", 180, "CloudWatch metric statistics period")
	cmd.Flags().StringVar(&statistic, "statistic", "average", "CloudWatch statistics method")
	cmd.Flags().StringVar(&dbClusterId, "db_cluster_id", "", "DB cluster identifier")
	cmd.Flags().BoolVar(&accpetNil, "accept_nil", false, "Continue if CloudWatch provides no metrics for the time period")
	cmd.Flags().StringVar(&availabilityZoneSeverity, "available_zone_severity", "critical", "Trigger a #{severity} if availability zone is different than given argument")
	cmd.Flags().StringVar(&availabilityZone, "available_zone", "us-west-1a", "available zone")
	cmd.Flags().Float64Var(&cpuCritical, "cpu_critical_over", 80, "Trigger a critical if cpu usage is over a percentage")
	cmd.Flags().Float64Var(&cpuWarning, "cpu_warning_over", 40, "Trigger a warning if cpu usage is over a percentage")
	cmd.Flags().Float64Var(&memoryCritical, "memory_critical_over", 80, "Trigger a critical if memory usage is over a Bytes")
	cmd.Flags().Float64Var(&memoryWarning, "memory_warning_over", 40, "Trigger a warning if memory usage is over a Bytes")
	cmd.Flags().Float64Var(&diskCritical, "disk_critical_over", 80, "Trigger a critical if disk usage is over a Bytes")
	cmd.Flags().Float64Var(&diskWarning, "disk_warning_over", 40, "Trigger a warning if disk usage is over a Bytes")
	cmd.Flags().Float64Var(&conectionCritical, "connections_critical_over", 80, "Trigger a critical if connection number is over a number")
	cmd.Flags().Float64Var(&connectionWarning, "connections_warning_over", 40, "Trigger a warning if connection number is over a number")
	cmd.Flags().Float64Var(&iopsCritical, "iops_critical_over", 80, "Trigger a critical if iops number is over a Count/Second")
	cmd.Flags().Float64Var(&iopsWarning, "iops_warning_over", 40, "Trigger a warning if connection number is over a Count/Second")
	cmd.Flags().StringVar(&roleArn, "role_arn", "", "AWS role arn of the role of the third party account to switch to")
	return cmd
}

func checkCPU(value *float64, instance string) {
	if checkNilValue(value, instance, "cpu") {
		return
	}

	if *value >= metricSeverities["CPUUtilization"]["critical"] {
		fmt.Println("CRITICAL : For DB Instance :", instance, "latest cpu usage value :", *value, "expected lower than", metricSeverities["CPUUtilization"]["critical"])
	}

	if *value >= metricSeverities["CPUUtilization"]["warning"] {
		fmt.Println("WARNING : For DB Instance :", instance, "latest cpu usage value :", *value, "expected lower than", metricSeverities["CPUUtilization"]["warning"])
	}
}

func checkMemory(value *float64, instance string, instanceClass string) {
	if checkNilValue(value, instance, "memory") {
		return
	}

	memoryTotalBytes := getMemoryTotalBytes(instanceClass)
	memoryUsageBytes := memoryTotalBytes - *value
	memoryUsagePercentage := (memoryUsageBytes / memoryTotalBytes) * 100

	if memoryUsagePercentage >= metricSeverities["FreeableMemory"]["critical"] {
		fmt.Println("CRITICAL : For DB Instance :", instance, "latest memory usage value :", memoryUsagePercentage, "expected lower than", metricSeverities["FreeableMemory"]["critical"])
	}

	if memoryUsagePercentage >= metricSeverities["FreeableMemory"]["warning"] {
		fmt.Println("WARNING : For DB Instance :", instance, "latest memory usage value :", memoryUsagePercentage, "expected lower than", metricSeverities["FreeableMemory"]["warning"])
	}
}

func checkDiskSpace(value *float64, instance string, allocatedStorage int64) {
	if checkNilValue(value, instance, "disk") {
		return
	}

	diskTotalBytes := float64(allocatedStorage) * math.Pow(1024, 3)
	diskUsageBytes := diskTotalBytes - *value
	diskUsagePercentage := (diskUsageBytes / diskTotalBytes) * 100

	if diskUsagePercentage >= metricSeverities["FreeStorageSpace"]["critical"] {
		fmt.Println("CRITICAL : For DB Instance :", instance, "latest disk usage value :", diskUsagePercentage, "expected lower than", metricSeverities["FreeStorageSpace"]["critical"])
	}

	if diskUsagePercentage >= metricSeverities["FreeStorageSpace"]["warning"] {
		fmt.Println("WARNING : For DB Instance :", instance, "latest disk usage value :", diskUsagePercentage, "expected lower than", metricSeverities["FreeStorageSpace"]["warning"])
	}
}

func checkConnections(value *float64, instance string) {
	if checkNilValue(value, instance, "database connections") {
		return
	}

	if *value >= metricSeverities["DatabaseConnections"]["critical"] {
		fmt.Println("CRITICAL : For DB Instance :", instance, "latest cpu usage value :", *value, "expected lower than", metricSeverities["DatabaseConnections"]["critical"])
	}

	if *value >= metricSeverities["DatabaseConnections"]["warning"] {
		fmt.Println("WARNING : For DB Instance :", instance, "latest cpu usage value :", *value, "expected lower than", metricSeverities["DatabaseConnections"]["warning"])
	}
}

func checkIops(value *float64, value2 *float64, instance string) {
	isReadIopsNull := checkNilValue(value, instance, "iops")

	if isReadIopsNull {
		return
	}

	isWriteIopsNull := checkNilValue(value2, instance, "iops")

	if isWriteIopsNull {
		return
	}

	iopsValue := *value + *value2
	if iopsValue >= metricSeverities["CPUUtilization"]["critical"] {
		fmt.Println("CRITICAL : For DB Instance :", instance, "latest iops usage value :", iopsValue, "expected lower than", metricSeverities["CPUUtilization"]["critical"])
	}

	if iopsValue >= metricSeverities["CPUUtilization"]["warning"] {
		fmt.Println("WARNING : For DB Instance :", instance, "latest iops usage value :", iopsValue, "expected lower than", metricSeverities["CPUUtilization"]["warning"])
	}
}

func checkNilValue(value *float64, instance string, metric string) bool {
	if value == nil && accpetNil {
		fmt.Println("DB INSTACE :", instance, ":", metric, "usage : Cloudwatch returned no results for time period. Accept nil passed so OK")
		return true
	}
	if value == nil && !accpetNil {
		fmt.Println("UNKNOWN : DB INSTACE :", instance, ":", metric, "usage : Requested time period did not return values from Cloudwatch. Try increasing your time period.")
		return true
	}
	return false
}

func getMemoryTotalBytes(instaceClass string) float64 {
	memoryByteMap := make(map[string]float64)
	memoryByteMap["db.cr1.8xlarge"] = 244.0
	memoryByteMap["db.m1.small"] = 1.7
	memoryByteMap["db.m1.medium"] = 3.75
	memoryByteMap["db.m1.large"] = 7.5
	memoryByteMap["db.m1.xlarge"] = 15.0
	memoryByteMap["db.m2.xlarge"] = 17.1
	memoryByteMap["db.m2.2xlarge"] = 34.2
	memoryByteMap["db.m2.4xlarge"] = 68.4
	memoryByteMap["db.m3.medium"] = 3.75
	memoryByteMap["db.m3.large"] = 7.5
	memoryByteMap["db.m3.xlarge"] = 15.0
	memoryByteMap["db.m3.2xlarge"] = 30.0
	memoryByteMap["db.m4.large"] = 8.0
	memoryByteMap["db.m4.xlarge"] = 16.0
	memoryByteMap["db.m4.2xlarge"] = 32.0
	memoryByteMap["db.m4.4xlarge"] = 64.0
	memoryByteMap["db.m4.10xlarge"] = 160.0
	memoryByteMap["db.m4.16xlarge"] = 256.0
	memoryByteMap["db.m5.large"] = 8.0
	memoryByteMap["db.m5.xlarge"] = 16.0
	memoryByteMap["db.m5.2xlarge"] = 32.0
	memoryByteMap["db.m5.4xlarge"] = 64.0
	memoryByteMap["db.m5.12xlarge"] = 192.0
	memoryByteMap["db.m5.24xlarge"] = 384.0
	memoryByteMap["db.r3.large"] = 15.0
	memoryByteMap["db.r3.xlarge"] = 30.5
	memoryByteMap["db.r3.2xlarge"] = 61.0
	memoryByteMap["db.r3.4xlarge"] = 122.0
	memoryByteMap["db.r3.8xlarge"] = 244.0
	memoryByteMap["db.r4.large"] = 15.25
	memoryByteMap["db.r4.xlarge"] = 30.5
	memoryByteMap["db.r4.2xlarge"] = 61.0
	memoryByteMap["db.r4.4xlarge"] = 122.0
	memoryByteMap["db.r4.8xlarge"] = 244.0
	memoryByteMap["db.r4.16xlarge"] = 488.0
	memoryByteMap["db.t1.micro"] = 0.615
	memoryByteMap["db.t2.micro"] = 1.0
	memoryByteMap["db.t2.small"] = 2.0
	memoryByteMap["db.t2.medium"] = 4.0
	memoryByteMap["db.t2.large"] = 8.0
	memoryByteMap["db.t2.xlarge"] = 16.0
	memoryByteMap["db.t2.2xlarge"] = 32.0
	memoryByteMap["db.x1.16xlarge"] = 976.0
	memoryByteMap["db.x1.32xlarge"] = 1952.0
	memoryByteMap["db.x1e.xlarge"] = 122.0
	memoryByteMap["db.x1e.2xlarge"] = 244.0
	memoryByteMap["db.x1e.4xlarge"] = 488.0
	memoryByteMap["db.x1e.8xlarge"] = 976.0
	memoryByteMap["db.x1e.16xlarge"] = 1952.0
	memoryByteMap["db.x1e.32xlarge"] = 3904.0

	return memoryByteMap[instaceClass] * math.Pow(1024, 3)
}

func getClusterDetails() error {
	if len(dbClusterId) > 0 {
		dbclustersInput := &rds.DescribeDBClustersInput{}

		filter := &rds.Filter{}
		filter.Name = aws.String("db-cluster-id")
		filter.Values = []*string{&dbClusterId}
		dbclustersInput.Filters = []*rds.Filter{filter}

		dbClusterOutput, err := rdsClient.DescribeDBClusters(dbclustersInput)

		if err != nil {
			fmt.Println("An error occurred processing AWS RDS API DescribeDBClusters", err)
			return err
		}

		if dbClusterOutput == nil || dbClusterOutput.DBClusters == nil || len(dbClusterOutput.DBClusters) <= 0 {
			fmt.Println("UNKNOWN : DB Cluster not found!")
		}

		if dbClusterOutput != nil && dbClusterOutput.DBClusters != nil && len(dbClusterOutput.DBClusters) > 0 {
			for _, dbclustersMember := range dbClusterOutput.DBClusters[0].DBClusterMembers {
				if *dbclustersMember.IsClusterWriter {
					dbInstanceInput := &rds.DescribeDBInstancesInput{}
					filter := &rds.Filter{}
					filter.Name = aws.String("db-instance-id")
					filter.Values = []*string{aws.String(*dbclustersMember.DBInstanceIdentifier)}
					dbInstanceInput.Filters = []*rds.Filter{filter}

					dbClusterOutput, err := rdsClient.DescribeDBInstances(dbInstanceInput)

					if err != nil {
						fmt.Println("An error occurred processing AWS RDS API DescribeDBInstances", err)
						return err
					}

					if dbClusterOutput == nil || dbClusterOutput.DBInstances == nil || len(dbClusterOutput.DBInstances) <= 0 {
						fmt.Println("UNKNOWN :", dbInstanceId, "instance not found")
					} else {
						dbInstanceZoneMapping[*dbclustersMember.DBInstanceIdentifier] = *dbClusterOutput.DBInstances[0].AvailabilityZone
						instanceClassMapping[*dbclustersMember.DBInstanceIdentifier] = *dbClusterOutput.DBInstances[0].DBInstanceClass
						allocatedStorageMapping[*dbclustersMember.DBInstanceIdentifier] = *dbClusterOutput.DBInstances[0].AllocatedStorage
					}
				}
			}
		}
	}
	return nil
}

func getDbInstanceDetails() error {
	if len(dbInstanceId) > 0 {
		dbInstanceInput := &rds.DescribeDBInstancesInput{}
		filter := &rds.Filter{}
		filter.Name = aws.String("db-instance-id")
		filter.Values = []*string{aws.String(dbInstanceId)}
		dbInstanceInput.Filters = []*rds.Filter{filter}

		dbClusterOutput, err := rdsClient.DescribeDBInstances(dbInstanceInput)

		if err != nil {
			fmt.Println("An error occurred processing AWS RDS API DescribeDBInstances", err)
			return err
		}

		if dbClusterOutput == nil || dbClusterOutput.DBInstances == nil || len(dbClusterOutput.DBInstances) <= 0 {
			fmt.Println("UNKNOWN :", dbInstanceId, "instance not found")
		} else {
			dbInstanceZoneMapping[dbInstanceId] = *dbClusterOutput.DBInstances[0].AvailabilityZone
			instanceClassMapping[dbInstanceId] = *dbClusterOutput.DBInstances[0].DBInstanceClass
			allocatedStorageMapping[dbInstanceId] = *dbClusterOutput.DBInstances[0].AllocatedStorage
		}
	}
	return nil
}

func main() {
	rootCmd := configureRootCommand()
	if err := rootCmd.Execute(); err != nil {
		fmt.Fprintf(os.Stderr, "error: %v\n", err)
		os.Exit(1)
	}
}

func run(cmd *cobra.Command, args []string) error {
	if len(args) != 0 {
		_ = cmd.Help()
		return fmt.Errorf("invalid argument(s) received")
	}
	checkRds()
	return nil
}
